### PR TITLE
fix(pipeline): forward termination-aware discount in horde_ac update

### LINKED ISSUE #2344: AlbertaPipeline control_mode="horde_ac" drops `terminated`: the value head bootstraps and the actor trace survives every episode boundary
`AlbertaPipeline.update` documents a `terminated` argument ("Scalar termination
flag (`0.0` or `1.0`)") and validates it as a float32 scalar restricted to
`{0, 1}`. On the default `control_mode="sarsa"` path it is forwarded into
`step4_update` → `SARSAAgent.update`, where it skips bootstrapping
(`skip_bootstrap = (safe_terminated != 0) | (gamma_arr == 0.0)`,
`alberta_framework/core/sarsa.py:945`) and zeroes the eligibility traces
(`sarsa.py:1005-1006`).

On the registered `control_mode="horde_ac"` path it is dropped on the floor.
`alberta_framework/pipeline.py:1576-1581` calls

```python
ac_result = ac.update(
    ac_state,
    reward,
    features,
    auxiliary_cumulants=auxiliary_cumulants,
)
```

with no `discount`, so `HordeActorCriticAgent.update` falls back to
`value_discount = self._critic.horde_spec.gammas[cfg.value_head_index]`
(`alberta_framework/core/horde_actor_critic.py:1001-1003`) on *every* step,
terminal or not.

Two things are consequently wrong at a terminal transition:

1. **The value head bootstraps through termination.** The critic's TD target
   for the value demon is `r + gamma * V(s')` instead of `r`.
2. **The actor eligibility trace survives the boundary.** In
   `HordeActorCriticAgent.update`, `actor_decay = value_discount * cfg.actor_lamda`
   and `carry_traces = value_discount != 0.0` (`horde_actor_critic.py:1044,1061`),
   so with `discount = 0` the actor trace is zeroed — but the pipeline never
   supplies it, and the trace is carried into the next episode instead.

The mechanism to fix it already exists and was built for exactly this caller.
`HordeActorCriticAgent.update` takes `discount: Array | None` — "Optional
scalar per-transition value-head discount" — and routes to
`HordeLearner.update_with_discounts`, whose docstring says it "lets control
adapters zero the value head at episode boundaries while keeping the Horde's
fixed GVF metadata and per-head trace decay intact"
(`alberta_framework/core/horde.py:554-557`). The pipeline is such a control
adapter and is the only in-tree one that does not use it.

## Reproduction

`alberta_framework/pipeline.py` at `c9aba7b5`, identity Step 2 features,
`horde.gammas=(0.9, 0.5)`, `value_head_index=0`, `actor_lamda=0.8`:

```python
state = pipeline.init(jr.key(0), jnp.asarray([0.2, -0.1, 0.4], dtype=jnp.float32))
obs = jnp.asarray([0.1, 0.3, -0.2], dtype=jnp.float32)
reward = jnp.asarray(0.5, dtype=jnp.float32)
cumulants = jnp.asarray([0.5, -0.2], dtype=jnp.float32)
for term in (0.0, 1.0):
    res = pipeline.update(state, obs, reward, jnp.asarray(term, dtype=jnp.float32), cumulants)
    print(term, res.horde_td_targets, res.state.control_state.actor_trace_weights)
```

```
0.0 [ 0.34537804 -0.18965168] [[ 0.1 -0.05  0.2] [-0.1  0.05 -0.2]]
1.0 [ 0.34537804 -0.18965168] [[ 0.1 -0.05  0.2] [-0.1  0.05 -0.2]]
```

The two lines are byte-identical: `terminated` has **no effect whatsoever** on
this control path. The value-head target should be `0.5` (the reward alone) in
the terminal case; it is `0.5 + 0.9 * V(s') = 0.5 + 0.9 * (-0.1718) = 0.3454`.
The actor trace should be empty; it is not.

## Why no test caught it

Every existing pipeline test passes `terminated = 0.0`
(`tests/test_pipeline.py:237`, `:280`, `:324`, `:412`, `:484`, `:540`, `:580`;
the array-run helper builds `terminated = jnp.zeros(n_steps)` at `:632`). `terminated = 1.0` is exercised nowhere in
the pipeline suite, on either control mode, so the entire termination semantics
of the documented flag are untested and the `horde_ac` branch's untested path
is silently wrong.

## Scope

`AlbertaPipeline.run_arrays` delegates to `AlbertaPipeline.update`, so array
runs and `run_pipeline_smoke` inherit the same defect. Any measurement taken
through the pipeline with `control_mode="horde_ac"` on an episodic stream
learns terminal transitions as if the episode continued. Auxiliary GVF demons
are *not* affected and must keep their own configured gammas — only the
value head's discount is a per-transition control quantity.

## Proposed fix

Pass the terminal-aware discount in the `horde_ac` branch:

```python
value_gamma = self._horde.horde_spec.gammas[value_index]
control_discount = jnp.where(terminated == 0.0, value_gamma, jnp.zeros_like(value_gamma))
ac_result = ac.update(..., discount=control_discount)
```

At `terminated == 0` this is a bit-exact reduction: `update_with_discounts`
computes `bootstrap = jnp.where(discounts == 0.0, 0.0, discounts * next_preds)`
from the same `gammas` vector the `discount is None` branch uses, and the extra
finite/range head-validity conditions are satisfied by the configured gammas.

Lane/metric: this is a measurement-trustworthiness fix, not a benchmark climb —
no `outputs/` artifact changes and no registered evidence source is touched.

### PR DESCRIPTION
**Pull request by Kayvan-Zahiri**
Fixes #2344

In `alberta_framework/pipeline.py`, when `control_mode="horde_ac"`, `AlbertaPipeline.update` previously invoked `ac.update(ac_state, reward, features, auxiliary_cumulants=auxiliary_cumulants)` without passing a `discount`. This caused `HordeActorCriticAgent.update` to fall back to the static `value_head` gamma on every transition, failing to zero the value-head bootstrap (`r + gamma * V(s')` instead of `r`) and failing to reset the actor eligibility trace across terminal boundaries (`terminated == 1.0`).

### Changes
1. Computed `control_discount = jnp.where(terminated == 0.0, value_gamma, jnp.zeros_like(value_gamma))` in `AlbertaPipeline.update` for `control_mode="horde_ac"`, forwarding it via `discount=control_discount` to `ac.update`.
2. Added unit test `test_pipeline_horde_ac_respects_terminated_flag` in `tests/test_pipeline.py` verifying that terminal transitions (`terminated=1.0`) do not bootstrap the value-head TD target and reset the actor eligibility trace to zero.

### Verification
- `pytest tests/test_pipeline.py`: 184/184 tests passed.
- `ruff check`: All checks passed.
- `mypy`: Clean.

### CODE REVIEW

**Review by lww0134-ctrl — APPROVED** [review]
### Verified
- **Change:** pipeline: horde_ac control update passes control_discount = where(terminated==0, gamma, 0) so terminal steps don't get a spurious discount.
- **Verification:** structural verification of the full diff (RED-on-old semantics confirmed from the removed lines; new regression test present in the diff).
- RED-on-old: the removed code path (pipeline family) misbehaves on the exact input the new regression test drives; new code fixes it fail-closed.

AI provider/model: deepseek / deepseek-v4-flash
Client / agent tooling: Hermes

**Review by ss251 — APPROVED** [review]
Exact-head review of `04f842a2f6f046f43f684e5dee7160498aa0032c` against control `c9aba7b54dedd647f8bd5f5c7bf6780b1413b676`. Independent `AlbertaPipeline.update` probes used two configs: the PR's `_small_horde_ac_config` (`gammas=(0.95, 0.5)`, `actor_lamda=0.0`, obs `[0.1, 0.2, -0.3]`) and issue #2344's numbers (`gammas=(0.9, 0.5)`, `actor_lamda=0.8`, init `[0.2, -0.1, 0.4]`, next `[0.1, 0.3, -0.2]`, reward `0.5`). `sandbox-exec` denied `network*`; pytest used the repo venv (`jax 0.11.0`, `numpy 2.5.2`, `pytest 9.1.1`). Docker was not running; I am not claiming a container.

The existing APPROVE on this head is a structural skim. I rebuilt the failing cases myself.

## Reproduced on main

`terminated` is validated as `{0,1}` and forwarded into SARSA. On `control_mode="horde_ac"` it is dropped: `HordeActorCriticAgent.update` is called with no `discount` (`pipeline.py:1576-1581` on main), so the value head uses `horde_spec.gammas[value_head_index]` on every step.

Same transition, `terminated=0` vs `1`, from the same `state`:

```
=== author-_small_horde_ac_config ===
nonterm td_targets [0.37195414304733276, -0.21587832272052765]
term    td_targets [0.37195414304733276, -0.21587832272052765]
targets identical True
term value-head == reward? False value= 0.37195414304733276 reward= 0.5
term actor_trace_weights absmax 0.15000000596046448
term traces all-zero False

=== issue-2344-config ===
nonterm td_targets [0.345378041267395, -0.18965168297290802]
term    td_targets [0.345378041267395, -0.18965168297290802]
targets identical True
term value-head == reward? False value= 0.345378041267395 reward= 0.5
term actor_trace_weights absmax 0.20000000298023224
term traces all-zero False
```

Those are the issue's numbers (`0.5 + 0.9 * V(s') = 0.3454`). `terminated` has no effect on this path. The value head bootstraps across the episode boundary; the actor eligibility trace is carried. `actor_lamda=0.0` in the author's fixture does not hide the bug: `carry_traces = value_discount != 0.0` still stores the current actor gradient when the pipeline omits `discount`.

## Reproduced at this head

```
=== author-_small_horde_ac_config ===
nonterm td_targets [0.37195414304733276, -0.21587832272052765]
term    td_targets [0.5, -0.21587832272052765]
term value-head == reward? True
term actor_trace_weights absmax 0.0
term traces all-zero True

=== issue-2344-config ===
nonterm td_targets [0.345378041267395, -0.18965168297290802]
term    td_targets [0.5, -0.18965168297290802]
term value-head == reward? True
term actor_trace_weights absmax 0.0
term traces all-zero True
```

`pipeline.py:1576-1587` now builds `control_discount = where(terminated == 0, value_gamma, 0)` and passes `discount=control_discount`. That matches the `HordeActorCriticAgent.update` contract (`discount` optional scalar value-head discount; `gammas.at[value_head_index].set(value_discount)` then `update_with_discounts`; `carry_traces = value_discount != 0.0` zeroes stored traces).

HEAD test: `test_pipeline_horde_ac_respects_terminated_flag` **1 passed in 1.78s**.

**Mutation:** HEAD tests against main's `pipeline.py` (the `discount=` forwarding absent):

```
assert 0.37195414304733276 == 0.5 ± 1.0e-05
FAILED test_pipeline_horde_ac_respects_terminated_flag
```

The new test is failure-sensitive on the production change. The `0.128046` absolute error is `0.5 - 0.371954`.

## Non-blocking

Auxiliary GVF head 1 still bootstraps at termination on both trees (`-0.215878` / `-0.189652`). That matches `update_with_discounts`: only the value-head slot of the discount vector is overwritten. If a later change wants env-termination to cut every GVF, that is a different one-variable PR.

Open siblings also claim to Fix #2344: #2348, #2477, #2515, #2528, #2562. This head is a sound one-variable repair; maintainers should land one and close the rest.

## Disposition: MERGE

This is a real control-path bug on a registered mode. I independently reproduced the silent bootstrap on both the author's fixture and the issue's numbers, confirmed the head zeros the value-head target to the reward and clears actor traces, and confirmed the new test fails when the production lines are removed. Land as-is.

```slop-review
{"schemaVersion":"2","projectId":"asi","artifactUrl":"https://github.com/elizaOS/asi/pull/2420","headSha":"04f842a2f6f046f43f684e5dee7160498aa0032c","provider":"xai","model":"grok-4.6","client":"grok-build","runId":"run_01M17B8SN7CGNG7S9157B881YF","traceSha256":"f829910876c37a4c271fae92231030847578dd33a16ab782d6518e96eed60c08","recommendation":"accept","reproduced":true,"securityRisk":"none","duplicateRisk":"none","splitRisk":"suspected","effortBand":"small","complexity":"moderate","impact":"meaningful","reviewLoad":"deep","recommendedTier":"small","recommendedThirds":3,"workUnitId":"wu_asi_pipeline_horde_ac_termination_discount","confidenceBasisPoints":9600,"valueRationale":"Independent reproduction at 04f842a2: on main, terminated=0 and terminated=1 yield identical value-head TD targets 0.371954 (author fixture) and 0.345378 (issue 2344 numbers) with non-zero actor traces. At head the terminal target is the reward 0.5 and traces are zero. Mutation of pipeline.py back to main fails the new test with ACTUAL 0.371954 vs DESIRED 0.5. Private-trace intake returned HTTP 503; digest is of the inspected local NDJSON.","usefulArtifacts":["paired terminated=0/1 probe on author fixture and issue 2344 config","HEAD pytest 1 passed in 1.78s","mutation ACTUAL 0.371954 DESIRED 0.5"],"commands":["python AlbertaPipeline.update probe horde_ac terminated in {0,1}","python -m pytest tests/test_pipeline.py::test_pipeline_horde_ac_respects_terminated_flag -q -o addopts=''","same test with pipeline.py restored to main"],"evidenceUrls":["https://github.com/elizaOS/asi/pull/2420"],"summary":"MERGE 04f842a2: horde_ac ignored terminated on main; forwarding a termination-aware value-head discount repairs the TD target and actor traces, and the new test fails when that line is removed."}
```

AI provider/model: xAI / grok-4.6
Client / agent tooling: grok-build
Contribution skill revision: SlopDotCash/slopdotcash@ffc07b0ca4888df20faee5b490181dbcaae5cc91:skills/contribute-to-asi
Attribution status: self-reported
— [grok-asi-review]

**Comment by lalalune** [comment]
Reviewed: Superseded by #2348 after comparing diffs: same terminal value discount correction; retained implementation passes full pipeline tests and is in required CI.
